src/store/options.js

```diff
@@ -1,37 +1,71 @@
+/**
+ * Ghostery Browser Extension
+ * https://www.ghostery.com/
+ *
+ * Copyright 2017-present Ghostery GmbH. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0
+ */
+
 import { store } from 'hybrids';
 
 const UPDATE_ACTION_NAME = 'updateOptions';
 
+export function isUpdateOptionsMessage(message) {
+  return message?.action === UPDATE_ACTION_NAME;
+}
+
+export const DNR_IDS = chrome.runtime
+  .getManifest()
+  .declarative_net_request.rule_resources.map((r) => r.id);
+
 const Options = {
   trackerWheelDisabled: false,
   wtmSerpReport: true,
+  dnr: DNR_IDS.reduce((all, rule) => ({ ...all, [rule]: true }), {}),
   [store.connect]: {
     async get() {
-      const storage = await chrome.storage.local.get(['options']);
-      return storage.options || {};
+      const { options = {} } = await chrome.storage.local.get(['options']);
+      return options;
+    },
+    async set(_, options) {
+      await chrome.storage.local.set({ options });
+      return options;
     },
-    async set(_, options, keys) {
-      const prevOptions = await this.get();
-      const nextOptions = {
-        ...prevOptions,
-        ...Object.fromEntries(keys.map((key) => [key, options[key]])),
-      };
+    async observe(id, options) {
+      // Send update message to another contexts (background page / panel / options)
+      chrome.runtime.sendMessage({
+        action: UPDATE_ACTION_NAME,
+      });
 
-      chrome.storage.local.set({ options: nextOptions });
-      chrome.runtime.sendMessage({ action: UPDATE_ACTION_NAME });
+      // Ensure that DNR rulesets are equal to those from options.
+      // eg. when web extension updates, the rulesets are reset
+      // to the value from the manifest.
 
-      return options;
+      const enabledRulesetIds =
+        await chrome.declarativeNetRequest.getEnabledRulesets();
+
+      DNR_IDS.forEach((rule) => {
+        const enabled = options.dnr[rule];
+        if (enabled !== enabledRulesetIds.includes(rule)) {
+          chrome.declarativeNetRequest.updateEnabledRulesets({
```

Comment:
Would it make sense to merge these single updates into a single call where all updates are batched? Looks like the "updateEnabledRulesets" API was designed to be used that way.

Reply:
Yes and no ;) Two arguments against it:

1. We don't wait for the result (no `await` - on purpose, the action is very slow, it can take a few seconds, so we just trigger change
2. Currently, our UI does not have the option to change more than one dnr list at once.

Reply:
Yes, it needs to recompute the rules on every change. I have not done tests with it myself, but I remember from some WECG meeting that it is quite costly. On Desktop, it is less a concern, but on slow Mobile devices, you'll notice. Well, if the batch update will by design one have one change, then we cannot batch, agreed. Otherwise, it should avoid overhead (to batch), as the browser can skip the computation of intermediate states that will not be used.

The batch operation could also be triggered without blocking, so you could keep that non-blocking semantic.

Reply:
I migrated to a batch version.